proof_of_stake/src/validation.rs

```diff
@@ -0,0 +1,1295 @@
+         got {2}"
+    )]
+    InvalidTotalVotingPowerChange(u64, VotingPowerDelta, VotingPowerDelta),
+}
+
+/// An update of PoS data.
+#[derive(Clone, Debug)]
+pub enum DataUpdate<Address, TokenAmount, TokenChange>
+where
+    Address: Display
+        + Debug
+        + Clone
+        + PartialEq
+        + Eq
+        + PartialOrd
+        + Ord
+        + Hash
+        + BorshDeserialize
+        + BorshSerialize,
+    TokenAmount: Clone
+        + Debug
+        + Default
+        + Eq
+        + Sub
+        + Add<Output = TokenAmount>
+        + AddAssign
+        + BorshDeserialize
+        + BorshSerialize,
+    TokenChange: Display
+        + Debug
+        + Default
+        + Clone
+        + Copy
+        + Add<Output = TokenChange>
+        + Sub<Output = TokenChange>
+        + From<TokenAmount>
+        + Into<i128>
+        + PartialEq
+        + Eq
+        + BorshDeserialize
+        + BorshSerialize,
+{
+    /// PoS account's balance update
+    Balance(Data<TokenAmount>),
+    /// Bond update
+    Bond {
+        /// Bond ID
+        id: BondId<Address>,
+        /// Bond prior and posterior state
+        data: Data<Bonds<TokenAmount>>,
+        /// List of slashes applied to the bond's validator
+        slashes: Slashes,
+    },
+    /// Unbond update
+    Unbond {
+        /// Unbond ID
+        id: BondId<Address>,
+        /// Unbond prior and posterior state
+        data: Data<Unbonds<TokenAmount>>,
+        /// List of slashes applied to the bond's validator
+        slashes: Slashes,
+    },
+    /// A validator update
+    Validator {
+        /// Validator's address
+        address: Address,
+        /// Validator's data update
+        update: ValidatorUpdate<Address, TokenChange>,
+    },
+    /// Validator set update
+    ValidatorSet(Data<ValidatorSets<Address>>),
+    /// Total voting power update
+    TotalVotingPower(Data<TotalVotingPowers>),
+}
+
+/// An update of a validator's data.
+#[derive(Clone, Debug)]
+pub enum ValidatorUpdate<Address, TokenChange>
+where
+    Address: Clone + Debug,
+    TokenChange: Display
+        + Debug
+        + Default
+        + Clone
+        + Copy
+        + Add<Output = TokenChange>
+        + Sub<Output = TokenChange>
+        + PartialEq
+        + Eq
+        + BorshDeserialize
+        + BorshSerialize,
+{
+    /// Staking reward address update
+    StakingRewardAddress(Data<Address>),
+    /// Total deltas update
+    TotalDeltas(Data<ValidatorTotalDeltas<TokenChange>>),
+    /// Voting power update
+    VotingPowerUpdate(Data<ValidatorVotingPowers>),
+}
+
+/// Data update with prior and posterior state.
+#[derive(Clone, Debug)]
+pub struct Data<T>
+where
+    T: Clone + Debug,
+{
+    /// State before the update
+    pub pre: Option<T>,
+    /// State after the update
+    pub post: Option<T>,
+}
+
+/// Validate the given list of PoS data `changes`. Returns empty list, if all
+/// the changes are valid.
+pub fn validate<Address, TokenAmount, TokenChange>(
```

Comment:
Sorry about this massive fn! We'll add plenty tests for the validation and refactor it after the MVP into something more readable. Also, some of the validation in which we're looking for differences in larger data structures will be simplified by changing their storage layout to be more fine-grained, so we see the diffs at lower level from the storage key changes.